File: Tests/PCNoShitaNoChikaramochiTests/CharacterStateTests.swift
import XCTest

@testable import PCNoShitaNoChikaramochi

final class CharacterStateTests: XCTestCase {
    func testFromUsageBuckets() {
        XCTAssertEqual(CharacterState.from(usage: 0), .veryEasy)
        XCTAssertEqual(CharacterState.from(usage: 19.9), .veryEasy)
        XCTAssertEqual(CharacterState.from(usage: 20), .easy)
        XCTAssertEqual(CharacterState.from(usage: 39.9), .easy)
        XCTAssertEqual(CharacterState.from(usage: 40), .normal)
        XCTAssertEqual(CharacterState.from(usage: 59.9), .normal)
        XCTAssertEqual(CharacterState.from(usage: 60), .heavy)
        XCTAssertEqual(CharacterState.from(usage: 79.9), .heavy)
        XCTAssertEqual(CharacterState.from(usage: 80), .struggling)
        XCTAssertEqual(CharacterState.from(usage: 89.9), .struggling)
        XCTAssertEqual(CharacterState.from(usage: 90), .crushed)
        XCTAssertEqual(CharacterState.from(usage: 100), .crushed)
    }

    func testFromUsageHandlesOutOfRange() {
        XCTAssertEqual(CharacterState.from(usage: -10), .veryEasy)
        XCTAssertEqual(CharacterState.from(usage: 999), .crushed)
    }

    func testCycleDurationAnimatedStates() {
        XCTAssertEqual(CharacterState.veryEasy.cycleDuration, 0.4)
        XCTAssertEqual(CharacterState.easy.cycleDuration, 0.9)
        XCTAssertEqual(CharacterState.normal.cycleDuration, 1.6)
        XCTAssertEqual(CharacterState.heavy.cycleDuration, 3.2)
    }

    func testCycleDurationStaticStates() {
        XCTAssertNil(CharacterState.struggling.cycleDuration)
        XCTAssertNil(CharacterState.crushed.cycleDuration)
    }

    func testCycleDurationsAreMonotonicallyIncreasing() {
        let animated: [CharacterState] = [.veryEasy, .easy, .normal, .heavy]
        let durations = animated.compactMap(\.cycleDuration)
        XCTAssertEqual(durations, durations.sorted())
        XCTAssertEqual(durations.count, animated.count)
    }

    func testLiftKind() {
        XCTAssertEqual(CharacterState.veryEasy.liftKind, .dumbbell)
        XCTAssertEqual(CharacterState.easy.liftKind, .dumbbell)
        XCTAssertEqual(CharacterState.normal.liftKind, .dumbbell)
        XCTAssertEqual(CharacterState.heavy.liftKind, .barbell)
        XCTAssertEqual(CharacterState.struggling.liftKind, .barbell)
        XCTAssertEqual(CharacterState.crushed.liftKind, .barbell)
    }

    func testEveryCaseHasNonEmptyLabel() {
        for state in CharacterState.allCases {
            XCTAssertFalse(state.label.isEmpty, "\(state) のラベルが空")
        }
    }

    func testLabelsAreUnique() {
        let labels = CharacterState.allCases.map(\.label)
        XCTAssertEqual(Set(labels).count, labels.count)
    }

    func testEveryCaseProducesTint() {
        // Color は等価比較が不安定なので、全ケースで生成できること（クラッシュしないこと）を確認。
        XCTAssertEqual(CharacterState.allCases.map(\.tint).count, CharacterState.allCases.count)
    }
}

File: Windows.Tests/CharacterStateTests.cs
using PCNoShitaNoChikaramochi;
using Xunit;

namespace PCNoShitaNoChikaramochi.Tests;

public class CharacterStateTests
{
    [Theory]
    [InlineData(0, CharacterState.VeryEasy)]
    [InlineData(19.9, CharacterState.VeryEasy)]
    [InlineData(20, CharacterState.Easy)]
    [InlineData(39.9, CharacterState.Easy)]
    [InlineData(40, CharacterState.Normal)]
    [InlineData(59.9, CharacterState.Normal)]
    [InlineData(60, CharacterState.Heavy)]
    [InlineData(79.9, CharacterState.Heavy)]
    [InlineData(80, CharacterState.Struggling)]
    [InlineData(89.9, CharacterState.Struggling)]
    [InlineData(90, CharacterState.Crushed)]
    [InlineData(100, CharacterState.Crushed)]
    public void From_MapsUsageToState(double usage, CharacterState expected)
    {
        Assert.Equal(expected, CharacterStateExtensions.From(usage));
    }

    [Theory]
    [InlineData(-10, CharacterState.VeryEasy)]
    [InlineData(999, CharacterState.Crushed)]
    public void From_HandlesOutOfRange(double usage, CharacterState expected)
    {
        Assert.Equal(expected, CharacterStateExtensions.From(usage));
    }

    [Theory]
    [InlineData(CharacterState.VeryEasy, 0.4)]
    [InlineData(CharacterState.Easy, 0.9)]
    [InlineData(CharacterState.Normal, 1.6)]
    [InlineData(CharacterState.Heavy, 3.2)]
    public void CycleDuration_AnimatedStates(CharacterState state, double expected)
    {
        Assert.Equal(expected, state.CycleDuration());
    }

    [Theory]
    [InlineData(CharacterState.Struggling)]
    [InlineData(CharacterState.Crushed)]
    public void CycleDuration_StaticStatesAreNull(CharacterState state)
    {
        Assert.Null(state.CycleDuration());
    }

    [Theory]
    [InlineData(CharacterState.VeryEasy, LiftKind.Dumbbell)]
    [InlineData(CharacterState.Easy, LiftKind.Dumbbell)]
    [InlineData(CharacterState.Normal, LiftKind.Dumbbell)]
    [InlineData(CharacterState.Heavy, LiftKind.Barbell)]
    [InlineData(CharacterState.Struggling, LiftKind.Barbell)]
    [InlineData(CharacterState.Crushed, LiftKind.Barbell)]
    public void Lift_ChoosesEquipment(CharacterState state, LiftKind expected)
    {
        Assert.Equal(expected, state.Lift());
    }

    [Fact]
    public void Label_EveryStateHasUniqueNonEmptyLabel()
    {
        var states = Enum.GetValues<CharacterState>();
        var labels = states.Select(s => s.Label()).ToList();
        Assert.All(labels, l => Assert.False(string.IsNullOrEmpty(l)));
        Assert.Equal(labels.Count, labels.Distinct().Count());
    }
}

public class CharacterAnimatorTests
{
    [Fact]
    public void DefaultState_IsNormal()
    {
        var animator = new CharacterAnimator();
        Assert.Equal(CharacterState.Normal, animator.State);
        Assert.Equal(0.0, animator.Phase);
    }

    [Fact]
    public void Update_ChangesState()
    {
        var animator = new CharacterAnimator();
        animator.Update(95);
        Assert.Equal(CharacterState.Crushed, animator.State);
        animator.Update(10);
        Assert.Equal(CharacterState.VeryEasy, animator.State);
    }

    [Fact]
    public void Tick_StaticStateForcesPhaseToZero()
    {
        var animator = new CharacterAnimator();
        animator.Update(85); // Struggling: CycleDuration == null
        animator.Tick();
        Assert.Equal(0.0, animator.Phase);
    }

    [Fact]
    public void Tick_AnimatedStatePhaseStaysInRange()
    {
        var animator = new CharacterAnimator();
        animator.Update(10); // VeryEasy
        for (var i = 0; i < 50; i++)
        {
            animator.Tick();
            Assert.InRange(animator.Phase, 0.0, 1.0);
        }
    }
}
